Ui

expez opened the pull request:
The terminals and the views are implemented in the composite pattern:
terminal.draw will draw the terminal ui by calling the draw function of each view.

The mailbox format is read from the configuration file and with some metaprogramming the correct mailbox type is instantiated and used.  Thus, it should be very easy to add support for another mailbox at present.

I initially wanted to name the Terminal class 'Display', but since termbox has support for capturing key events I think it makes sense to go with this abstraction instead. Even though this leads to the Terminal class having more than one responsibility.  The future will thus bring Terminal.get_key_event, using the tb_poll_event method, which is a blocking call returning an event.  We can then use this event to find out what commands the user issued.

I found a nice dataset, of thousands of emails, used as a training set for machine learning systems tackling automatic spam detection.  The test mail included in the fixture is of somewhat explicit nature, and I expect more to follow :)

If you run this now it will output:

list of maildir folders | the email 
------------------------| itself
list of maildirlabels   | is here

I would like to be, eventually:

| folders | a list of emails in the active folder |
| --- | --- |
| labels | the currently active email |

Right now the size and position of the view is hardcoded in but it needs to be determined dynamically at runtime based on:
-The size of the terminal
-The length of the folder/label names

With tiling window managers being as popular as they are we should recheck the size of the terminal on every draw (termbox has a method for this)

expez committed:
Some project level refactorings.

- mv .hatterrc -> hatterc
- Add gitignore file
- Add rb_termbox to the Gemfile

expez committed:
Make configuration a singleton and add new settings.

expez committed:
Add test fixture with a single spam mail

expez committed:
Implement the views we want to display to the user.

The Folder_view is a list of all the folders of the mailbox.

Label_view is a list of all the labels of the mailbox.

The Mail_view displays the mail itself.

expez committed:
Implement maildir mailbox

This a stub for the first mailbox format.

To add support for another mailbox implement it in <maildir_format>_mailbox.
When the user adds maildir_format = <maildir_format> the file will be required
and the new mailbox used automagically.

expez committed:
Update rvmrc to set RUBYLIB to this project's path.

expez committed:
Implement terminal interface

jfredett commented:
I'm going to dig into this this weekend, I've got three days off to look at it.

jfredett commented on lib/mailbox/maildir_mailbox.rb:
```diff
@@ -0,0 +1,24 @@
+class Maildir_mailbox
```
Typically classes are named in full camel-case, eg "MaildirMailbox" would be more correct/idiomatic here.

jfredett commented on lib/views/folder_view.rb:
```diff
@@ -0,0 +1,22 @@
+class Folder_view < View
```
Same comment about naming as above.

jfredett commented on lib/views/label_view.rb:
```diff
@@ -0,0 +1,21 @@
+class Label_view < View
+
+  def initialize location
+    super
```
By default, ruby will proxy up to it's superclass, so this method impl isn't necessary.

jfredett commented on lib/views/label_view.rb:
```diff
@@ -0,0 +1,21 @@
+class Label_view < View
+
+  def initialize location
```
Defs should have parens around arguments. The github styleguide for ruby is probably worth reading here. It's not so much prescriptive as descriptive of what rubyists _do_, rather than what they "should do"

jfredett commented on lib/views/view.rb:
```diff
@@ -0,0 +1,73 @@
+class View
+  require 'configuration'
```
Requires should go outside of class scopes. They don't always follow the scoping rules like you think (sometimes `require` works like `load`, sometimes it doesn't). Safest and most idiomatic place is either at the top of the file, or occasionally (to resolve potentially circular dependencies) at the bottom. Though the latter is typically not necessary, since `require` caches/memoizes

jfredett commented on lib/views/view.rb:
```diff
@@ -0,0 +1,73 @@
+
+  def initialize location
+    @Border_width = 1
+    @start_x = location[:x0]
```
There is a lovely module in the standard library called 'forwardable' which provides a new peice of class syntax (actually just a normal 'class' method), called `#delegate`, which essentially writes the boilerplate `def foo; @ivar.foo; end` method for you. I don't know how possible it is to wrap up a location in a class, but between forwardable and that, you could wrap up these four lines as `delegate [:start_x, :start_y, :end_x, :end_y] => :location`, instead of spreading them on four lines.

jfredett commented on lib/views/view.rb:
```diff
@@ -0,0 +1,73 @@
+  attr_reader :colors, :mailbox
+
+  def initialize location
+    @Border_width = 1
```
ivars (every var, actually) should be lower case. The only things with capitals are typically constants of some flavor.

jfredett commented on lib/views/view.rb:
```diff
@@ -0,0 +1,73 @@
+    set_colors
+  end
+
+  def draw_text text, x, y
```
Nice, that looks like the ncurses/terminal UI code I remember. :)

jfredett commented on spec/fixtures/maildir/INBOX/cur/TRAIN_00001.eml:
```diff
@@ -0,0 +1,63 @@
+From lkqyvstyles@site-personals.com  Tue Aug 27 03:56:47 2002
```
This is not really an appropriate email to use as a fixture... Can we make it a little more... SFW?

jfredett commented on spec/views/view_spec.rb:
```diff
@@ -0,0 +1,21 @@
+require 'spec_helper'
```
In general I'd like to see some more tests here, but I imagine poking the UI is probably difficult, so I think it's probably okay here.

jfredett commented:
I want to look at it some more, but I put a bunch of comments in on style stuff. Also, the fake email should be changed for something a bit less risque.

expez committed:
Refactor to be in line with the github style guide.

- Move require statements outside class defs.
- Use parens around defs with args.
- Remove implicit calls to super
- Replace offensive test fixture email.

jfredett commented on lib/configuration.rb:
```diff
@@ -1,18 +1,25 @@
 
+  class Colors < ConfigToolkit::BaseConfig
+  add_required_param(:foreground, String)
+  add_required_param(:background, String)
```
These lines should be indented 2 spaces (not a blocker, but if you fix it I'll love you forever)

jfredett commented:
Lovely, This is a great start. Merged.
